go/enclave/txpool/txpool.go

```diff
@@ -33,30 +34,69 @@ type TxPool struct {
 	running      bool
 	stateMutex   sync.Mutex
 	logger       gethlog.Logger
+	validateOnly atomic.Bool
 }
 
 // NewTxPool returns a new instance of the tx pool
-func NewTxPool(blockchain *ethchainadapter.EthChainAdapter, gasTip *big.Int, logger gethlog.Logger) (*TxPool, error) {
+func NewTxPool(blockchain *ethchainadapter.EthChainAdapter, gasTip *big.Int, validateOnly bool, logger gethlog.Logger) (*TxPool, error) {
 	txPoolConfig := ethchainadapter.NewLegacyPoolConfig()
 	legacyPool := legacypool.New(txPoolConfig, blockchain)
 
-	return &TxPool{
+	txp := &TxPool{
 		Chain:        blockchain,
 		txPoolConfig: txPoolConfig,
 		legacyPool:   legacyPool,
 		gasTip:       gasTip,
 		stateMutex:   sync.Mutex{},
+		validateOnly: atomic.Bool{},
 		logger:       logger,
-	}, nil
+	}
+	txp.validateOnly.Store(validateOnly)
+	go txp.start()
+	return txp, nil
+}
+
+func (t *TxPool) ChangeMode(validateOnly bool) {
+	t.validateOnly.Store(validateOnly)
 }
 
-// Start starts the pool
 // can only be started after t.blockchain has at least one block inside
-func (t *TxPool) Start() error {
+func (t *TxPool) start() {
 	if t.running {
-		return fmt.Errorf("tx pool already started")
+		return
 	}
 
+	cb := t.Chain.CurrentBlock()
+	if cb != nil && cb.Number.Uint64() > common.L2GenesisHeight+1 {
+		err := t._startInternalPool()
+		if err != nil {
+			t.logger.Crit("Failed to start tx pool", log.ErrKey, err)
+		}
+		return
+	}
+
+	var (
+		newHeadCh  = make(chan core.ChainHeadEvent)
+		newHeadSub = t.Chain.SubscribeChainHeadEvent(newHeadCh)
+	)
+	defer newHeadSub.Unsubscribe()
+	defer close(newHeadCh)
+	for { //nolint:gosimple
+		select {
+		case event := <-newHeadCh:
```

Comment:
I think this could hang forever and leave a stuck go-routine if something goes wrong. Maybe worth adding a timeout case as well, like `case <-time.After(someTimeout):`?